Add Emacs shortcuts

RohanM opened the pull request:
Hello!

I've made some additions in this PR to add support for some common Emacs shortcuts (without adding any default hotkeys). Would that be something you'd be open to merging into the plugin?

I'm fairly new to CodeMirror, so I'm sure there are better ways to do everything - let me know if you see anything that could be simpler. Thank you!

timhor commented:
Sure, happy to take a look!

timhor commented on src/__tests__/actions.spec.ts:
```diff
@@ -110,6 +113,30 @@ describe('Code Editor Shortcuts: actions', () => {
+    });
+
+    it('should delete the newline when at the end of the line', () => {
+      editor.setCursor({ line: 1, ch: 10 });
```
Should this be `ch: 9` to more accurately capture 'end of line'? (Since position 10 is past the end of the text)

```
dolor sit
0123456789
```

RohanM replied:
That's absolutely right, I've fixed that up.

RohanM commented:
Thanks for having a look through! I've added some specs for the behaviour when there's a selection and switched to using the natural cursor location instead of `getCursor('from|to')` in a few spots.

timhor commented:
Using `Go to next line` when on the last line unexpectedly breaks the sentence into two. Does this happen for you too? If so, can you add another test case together with your fix, that it `'should not navigate past the end of the document'`?

![next-line-bug](https://user-images.githubusercontent.com/14193387/157589999-82e5fbc1-347d-4895-9292-60d82bef108f.gif)

Also, could you please document the new behaviours in the table inside README.md as well?

timhor commented on src/actions.ts:
```diff
@@ -123,6 +136,45 @@ export const goToLineBoundary = (editor: Editor, boundary: 'start' | 'end') => {
 
+export const navigateLine = (editor: Editor, direction: 'up' | 'down') => {
+  const pos = editor.getCursor();
+  let line;
```
_nit_: would prefer adding a type if it's not initialised, i.e. `let line: number`

RohanM commented:
> Using Go to next line when on the last line unexpectedly breaks the sentence into two. Does this happen for you too?

I couldn't reproduce this one on Linux or MacOS (the cursor just goes to the end of the line without modifying the text). However, there's definitely a missing guard there, so I've added that in. Does that fix the bug for you?

timhor commented on README.md:
```diff
@@ -1,20 +1,25 @@
 # Code Editor Shortcuts
 
-This [Obsidian](https://obsidian.md) plugin adds keyboard shortcuts (hotkeys) commonly found in code editors such as Visual Studio Code or Sublime Text.
+This [Obsidian](https://obsidian.md) plugin adds keyboard shortcuts (hotkeys) commonly found in code editors such as Visual Studio Code, Sublime Text, Emacs or Vim.
```
I'll change this back for now because while I'm happy to have these few shortcuts added, I'm not prepared to support an additional two editors' worth of shortcuts (I have my hands full with existing feature requests as it is). And Obsidian already has a built-in option to enable Vim key bindings anyway.

timhor commented on src/__tests__/actions.spec.ts:
```diff
@@ -330,6 +465,13 @@ describe('Code Editor Shortcuts: actions', () => {
       expect(cursor.line).toEqual(0);
     });
 
+    it('should delete to the end of the line', () => {
+      deleteToEndOfLine(editor as any);
+
+      const { doc } = getDocumentAndSelection(editor);
+      expect(doc).toEqual('lorem ipsum\ndolor\namet');
+    });
```
Is this a duplicate of https://github.com/RohanM/obsidian-emacs-shortcuts/blob/add-emacs-shortcuts/src/__tests__/actions.spec.ts#L116-L122?

RohanM replied:
This spec starts with a range selection, so slightly different starting conditions.

timhor replied:
Ah of course! Thanks for clarifying 🙂

timhor commented:
> However, there's definitely a missing guard there, so I've added that in. Does that fix the bug for you?

Yes, that does fix the bug 👍

timhor commented:
Thanks for the PR!

RohanM commented:
No worries, thanks for your help pulling it together!
